2.5-huffman.ipynb

```diff
@@ -1,25 +1,208 @@
-    "   \n",
-    "* Indicates a higher likelihood of execution\n",
-    "* Means the script will be placed closer to the root if possible\n",
-    "* Results in a smaller inclusion proof and lower spending fees"
+    "* A higher likelihood of execution\n",
+    "* A position which is closer to the root (if possible)\n",
+    "* A smaller inclusion proof"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 1: Huffman Algorithm\n",
+    "\n",
+    "The Huffman algorithm to construct a TapTree structure can be described as follows:\n",
+    "\n",
+    "* Place all leafs in a node queue and sort by weight*\n",
+    "* While length ( node queue ) > 1 :\n",
+    "    * Pop the two elements with the highest weights\n",
+    "    * Generate parent node from popped elements\n",
+    "    * Add parent node to queue and resort\n",
+    "* Root node is represented by final queue element"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![test](images/huffman_intro0.jpg)"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "![test](images/taptree3.jpg)"
+    "**Note:** If more than 2 leafs or nodes share the same weight during construction with the Huffman algorithm, the selection of the two nodes to generate the next parent becomes ambiguous. In our implementation, we first sort our node queue by `weight`, then by `tagged hash` value. This way, the sorting of the node queue is always unambiguous given a distinct set of leafs."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 2.5.1: Construct a TapTree with the Huffman algorithm\n",
+    "\n",
+    "* We manually construct the TapTree from a set of 5 pay-to-pubkey TapLeafs with the same weights as shown in the image above."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "internal_pubkey = ECPubKey()\n",
+    "internal_pubkey.set(bytes.fromhex('03af455f4989d122e9185f8c351dbaecd13adca3eef8a9d38ef8ffed6867e342e3'))\n",
+    "\n",
+    "# Derive pay-to-pubkey tapleafs\n",
+    "privkeyA, pubkeyA = generate_key_pair()\n",
+    "privkeyB, pubkeyB = generate_key_pair()\n",
+    "privkeyC, pubkeyC = generate_key_pair()\n",
+    "privkeyD, pubkeyD = generate_key_pair()\n",
+    "privkeyE, pubkeyE = generate_key_pair()\n",
+    "\n",
+    "tapleafA = TapLeaf().construct_pk(pubkeyA)\n",
+    "tapleafB = TapLeaf().construct_pk(pubkeyB)\n",
+    "tapleafC = TapLeaf().construct_pk(pubkeyC)\n",
+    "tapleafD = TapLeaf().construct_pk(pubkeyD)\n",
+    "tapleafE = TapLeaf().construct_pk(pubkeyE)\n",
+    "\n",
+    "# Sorted queue: (5, A), (4, B), (3, C), (2, D), (1, E)\n",
```

Comment:
Confused how the sorting and selecting works (see above), so I cannot comment further until I 'get it'